{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os,sys,glob\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from utils import util"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### load rcsb pdb data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('table.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Id', 'Element', 'Ion', 'Binding family',\n",
       "       'Amino acids or nucleotide residues names', 'All bound residues',\n",
       "       'Is homomer', 'Number of bound amino acids or nucleotide residues',\n",
       "       'Number of all bound residues', 'Number of coordinating chains',\n",
       "       'Representative'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_homo = df[(df['Is homomer'] == 1) & ~(df['Amino acids or nucleotide residues names'].str.contains('DG')) & ~(df['Amino acids or nucleotide residues names'].str.contains('DT')) & ~(df['Amino acids or nucleotide residues names'].str.contains('DC')) & ~(df['Amino acids or nucleotide residues names'].str.contains('DA'))]#.tail(40)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_homo['Amino acids or nucleotide residues names'].str.contains('DU').sum()\n",
    "#'DG' in df['Amino acids or nucleotide residues names']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Id</th>\n",
       "      <th>Element</th>\n",
       "      <th>Ion</th>\n",
       "      <th>Binding family</th>\n",
       "      <th>Amino acids or nucleotide residues names</th>\n",
       "      <th>All bound residues</th>\n",
       "      <th>Is homomer</th>\n",
       "      <th>Number of bound amino acids or nucleotide residues</th>\n",
       "      <th>Number of all bound residues</th>\n",
       "      <th>Number of coordinating chains</th>\n",
       "      <th>Representative</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>1NC7-MG-1</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>T4</td>\n",
       "      <td>THR.</td>\n",
       "      <td>THR.</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>1NC7-MG-2</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>L4</td>\n",
       "      <td>LEU.</td>\n",
       "      <td>LEU.</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>1NC7-MG-3</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>T4</td>\n",
       "      <td>THR.</td>\n",
       "      <td>HOH.</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1041</th>\n",
       "      <td>4GGH-MG-1</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>N4</td>\n",
       "      <td>ASN.</td>\n",
       "      <td>ASN.HOH.</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1042</th>\n",
       "      <td>4GGH-MG-2</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>N4</td>\n",
       "      <td>ASN.</td>\n",
       "      <td>ASN.HOH.</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21717</th>\n",
       "      <td>7WM2-K-2</td>\n",
       "      <td>K</td>\n",
       "      <td>K⁺</td>\n",
       "      <td>T4</td>\n",
       "      <td>THR.</td>\n",
       "      <td>THR.</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21718</th>\n",
       "      <td>7WM2-K-3</td>\n",
       "      <td>K</td>\n",
       "      <td>K⁺</td>\n",
       "      <td>G4</td>\n",
       "      <td>GLY.</td>\n",
       "      <td>GLY.</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21764</th>\n",
       "      <td>8A35-RB-1</td>\n",
       "      <td>RB</td>\n",
       "      <td>Rb⁺</td>\n",
       "      <td>T4</td>\n",
       "      <td>THR.</td>\n",
       "      <td>THR.</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21765</th>\n",
       "      <td>8A35-RB-2</td>\n",
       "      <td>RB</td>\n",
       "      <td>Rb⁺</td>\n",
       "      <td>T4</td>\n",
       "      <td>THR.</td>\n",
       "      <td>THR.</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21766</th>\n",
       "      <td>8A35-RB-3</td>\n",
       "      <td>RB</td>\n",
       "      <td>Rb⁺</td>\n",
       "      <td>T4</td>\n",
       "      <td>THR.</td>\n",
       "      <td>THR.</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>589 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Id Element   Ion Binding family  \\\n",
       "39     1NC7-MG-1      MG  Mg²⁺             T4   \n",
       "40     1NC7-MG-2      MG  Mg²⁺             L4   \n",
       "41     1NC7-MG-3      MG  Mg²⁺             T4   \n",
       "1041   4GGH-MG-1      MG  Mg²⁺             N4   \n",
       "1042   4GGH-MG-2      MG  Mg²⁺             N4   \n",
       "...          ...     ...   ...            ...   \n",
       "21717   7WM2-K-2       K    K⁺             T4   \n",
       "21718   7WM2-K-3       K    K⁺             G4   \n",
       "21764  8A35-RB-1      RB   Rb⁺             T4   \n",
       "21765  8A35-RB-2      RB   Rb⁺             T4   \n",
       "21766  8A35-RB-3      RB   Rb⁺             T4   \n",
       "\n",
       "      Amino acids or nucleotide residues names All bound residues  Is homomer  \\\n",
       "39                                        THR.               THR.        True   \n",
       "40                                        LEU.               LEU.        True   \n",
       "41                                        THR.               HOH.        True   \n",
       "1041                                      ASN.           ASN.HOH.        True   \n",
       "1042                                      ASN.           ASN.HOH.        True   \n",
       "...                                        ...                ...         ...   \n",
       "21717                                     THR.               THR.        True   \n",
       "21718                                     GLY.               GLY.        True   \n",
       "21764                                     THR.               THR.        True   \n",
       "21765                                     THR.               THR.        True   \n",
       "21766                                     THR.               THR.        True   \n",
       "\n",
       "       Number of bound amino acids or nucleotide residues  \\\n",
       "39                                                     4    \n",
       "40                                                     4    \n",
       "41                                                     4    \n",
       "1041                                                   4    \n",
       "1042                                                   4    \n",
       "...                                                  ...    \n",
       "21717                                                  4    \n",
       "21718                                                  4    \n",
       "21764                                                  4    \n",
       "21765                                                  4    \n",
       "21766                                                  4    \n",
       "\n",
       "       Number of all bound residues  Number of coordinating chains  \\\n",
       "39                                4                              4   \n",
       "40                                4                              4   \n",
       "41                                2                              4   \n",
       "1041                              6                              4   \n",
       "1042                              6                              4   \n",
       "...                             ...                            ...   \n",
       "21717                             4                              4   \n",
       "21718                             4                              4   \n",
       "21764                             4                              4   \n",
       "21765                             4                              4   \n",
       "21766                             4                              4   \n",
       "\n",
       "       Representative  \n",
       "39               True  \n",
       "40               True  \n",
       "41               True  \n",
       "1041             True  \n",
       "1042            False  \n",
       "...               ...  \n",
       "21717            True  \n",
       "21718           False  \n",
       "21764           False  \n",
       "21765           False  \n",
       "21766           False  \n",
       "\n",
       "[589 rows x 11 columns]"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_homo[df_homo['Number of coordinating chains']==4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3732043/988947585.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_homo['pdb_id'] = df_homo['Id'].str.split('-', expand=True)[0]\n"
     ]
    }
   ],
   "source": [
    "df_homo['pdb_id'] = df_homo['Id'].str.split('-', expand=True)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Id</th>\n",
       "      <th>Element</th>\n",
       "      <th>Ion</th>\n",
       "      <th>Binding family</th>\n",
       "      <th>Amino acids or nucleotide residues names</th>\n",
       "      <th>All bound residues</th>\n",
       "      <th>Is homomer</th>\n",
       "      <th>Number of bound amino acids or nucleotide residues</th>\n",
       "      <th>Number of all bound residues</th>\n",
       "      <th>Number of coordinating chains</th>\n",
       "      <th>Representative</th>\n",
       "      <th>pdb_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5LDB-MG-1</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>T2</td>\n",
       "      <td>THR.</td>\n",
       "      <td>HOH.THR.</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>5LDB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2SHK-MG-1</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>A1G2R2</td>\n",
       "      <td>ALA.ARG.GLY.</td>\n",
       "      <td>ALA.GLN.HOH.</td>\n",
       "      <td>True</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>2SHK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2R6T-MG-1</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>D1I1</td>\n",
       "      <td>ASP.ILE.</td>\n",
       "      <td>ASP.ATP.ILE.</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>2R6T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2R6T-MG-2</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>D1I1</td>\n",
       "      <td>ASP.ILE.</td>\n",
       "      <td>ASP.ATP.ILE.</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>2R6T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2R6T-MG-3</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>D1I1</td>\n",
       "      <td>ASP.ILE.</td>\n",
       "      <td>ASP.ATP.ILE.</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>2R6T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2R6X-MG-1</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>D1I1</td>\n",
       "      <td>ASP.ILE.</td>\n",
       "      <td>ASP.ATP.HOH.ILE.</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>2R6X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>2R6X-MG-2</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>D1I1</td>\n",
       "      <td>ASP.ILE.</td>\n",
       "      <td>ASP.ATP.HOH.ILE.</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>2R6X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>2R6X-MG-3</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>D1I1</td>\n",
       "      <td>ASP.ILE.</td>\n",
       "      <td>ASP.ATP.HOH.ILE.</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>2R6X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2R8D-MG-1</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>D3</td>\n",
       "      <td>ASP.</td>\n",
       "      <td>ASP.HOH.</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>2R8D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>2R8D-MG-2</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>D3</td>\n",
       "      <td>ASP.</td>\n",
       "      <td>ASP.HOH.</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>2R8D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>2PLS-MG-1</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>D3</td>\n",
       "      <td>ASP.</td>\n",
       "      <td>ASP.HOH.</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>2PLS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>2PLS-MG-2</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>D3</td>\n",
       "      <td>ASP.</td>\n",
       "      <td>ASP.HOH.</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>2PLS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>2PLS-MG-3</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>D6</td>\n",
       "      <td>ASP.</td>\n",
       "      <td>ASP.</td>\n",
       "      <td>True</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>2PLS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>2Q0Y-MG-1</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>E2</td>\n",
       "      <td>GLU.</td>\n",
       "      <td>GLU.HOH.</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>2Q0Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>2Q0Y-MG-2</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>E2</td>\n",
       "      <td>GLU.</td>\n",
       "      <td>GLU.HOH.</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>2Q0Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>2P7V-MG-1</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>E2R2</td>\n",
       "      <td>ARG.GLU.</td>\n",
       "      <td>ARG.GLU.HOH.</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>2P7V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>2OX6-MG-1</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>E2</td>\n",
       "      <td>GLU.</td>\n",
       "      <td>GLU.HOH.</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>2OX6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>2OX6-MG-2</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>E2</td>\n",
       "      <td>GLU.</td>\n",
       "      <td>GLU.HOH.</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>2OX6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>2OX6-MG-3</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>E2</td>\n",
       "      <td>GLU.</td>\n",
       "      <td>GLU.</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>2OX6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>2PAN-MG-1</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>E2F2</td>\n",
       "      <td>GLU.PHE.</td>\n",
       "      <td>GLU.PHE.</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>2PAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>2PAN-MG-2</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>E2F2</td>\n",
       "      <td>GLU.PHE.</td>\n",
       "      <td>GLU.PHE.</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>2PAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>2PC5-MG-1</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>E3</td>\n",
       "      <td>GLU.</td>\n",
       "      <td>GLU.</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>2PC5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>2OI6-MG-1</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>D3</td>\n",
       "      <td>ASP.</td>\n",
       "      <td>ASP.HOH.</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>2OI6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>2OIW-MG-1</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>S2T2</td>\n",
       "      <td>SER.THR.</td>\n",
       "      <td>HOH.SER.THR.</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>2OIW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>2OIW-MG-2</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>S2T2</td>\n",
       "      <td>SER.THR.</td>\n",
       "      <td>HOH.SER.THR.</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>2OIW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>2IEC-MG-1</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>C2</td>\n",
       "      <td>CYS.</td>\n",
       "      <td>CYS.HOH.</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>2IEC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>2IEC-MG-2</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>C2</td>\n",
       "      <td>CYS.</td>\n",
       "      <td>CYS.HOH.</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>2IEC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>2IOU-MG-1</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>E3</td>\n",
       "      <td>GLU.</td>\n",
       "      <td>GLU.</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>2IOU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>2IOU-MG-2</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>E3F3</td>\n",
       "      <td>GLU.PHE.</td>\n",
       "      <td>GLU.PHE.</td>\n",
       "      <td>True</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>2IOU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>2IOU-MG-3</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>E3F3</td>\n",
       "      <td>GLU.PHE.</td>\n",
       "      <td>GLU.PHE.</td>\n",
       "      <td>True</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>2IOU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>2IOU-MG-4</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>E3</td>\n",
       "      <td>GLU.</td>\n",
       "      <td>GLU.</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>2IOU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>1NC7-MG-1</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>T4</td>\n",
       "      <td>THR.</td>\n",
       "      <td>THR.</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "      <td>1NC7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>1NC7-MG-2</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>L4</td>\n",
       "      <td>LEU.</td>\n",
       "      <td>LEU.</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "      <td>1NC7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>1NC7-MG-3</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>T4</td>\n",
       "      <td>THR.</td>\n",
       "      <td>HOH.</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "      <td>1NC7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>1Q5H-MG-1</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>E3</td>\n",
       "      <td>GLU.</td>\n",
       "      <td>GLU.HOH.</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "      <td>1Q5H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>1Q6Z-MG-1</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>L2N2R2</td>\n",
       "      <td>ARG.ASN.LEU.</td>\n",
       "      <td>ARG.ASN.LEU.</td>\n",
       "      <td>True</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>1Q6Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>1Q6Z-MG-2</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>L2N2R2</td>\n",
       "      <td>ARG.ASN.LEU.</td>\n",
       "      <td>ARG.ASN.LEU.</td>\n",
       "      <td>True</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>1Q6Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>1Q9W-MG-1</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>E2</td>\n",
       "      <td>GLU.</td>\n",
       "      <td>GLU.HOH.</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>1Q9W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>1QH8-MG-1</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>D2E1K1</td>\n",
       "      <td>ASP.GLU.LYS.</td>\n",
       "      <td>HOH.</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>1QH8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>1QH8-MG-2</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>D2E1K1</td>\n",
       "      <td>ASP.GLU.LYS.</td>\n",
       "      <td>ASP.GLU.HOH.LYS.</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>1QH8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>1QGU-MG-1</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>D2E1K1</td>\n",
       "      <td>ASP.GLU.LYS.</td>\n",
       "      <td>HOH.</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>1QGU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>1QGU-MG-2</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>D2E1K1</td>\n",
       "      <td>ASP.GLU.LYS.</td>\n",
       "      <td>ASP.GLU.HOH.LYS.</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>1QGU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>1QH1-MG-1</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>D2E1K1</td>\n",
       "      <td>ASP.GLU.LYS.</td>\n",
       "      <td>HOH.</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>1QH1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>1QH1-MG-2</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>D2E1K1</td>\n",
       "      <td>ASP.GLU.LYS.</td>\n",
       "      <td>ASP.GLU.HOH.LYS.</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>1QH1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>2HFK-MG-1</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>D2</td>\n",
       "      <td>ASP.</td>\n",
       "      <td>ASP.HOH.</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>2HFK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>2H7X-MG-1</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>D2</td>\n",
       "      <td>ASP.</td>\n",
       "      <td>ASP.HOH.</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>2H7X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>2H7Y-MG-1</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>D2</td>\n",
       "      <td>ASP.</td>\n",
       "      <td>ASP.HOH.</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>2H7Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>2A5L-MG-1</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>K2S2</td>\n",
       "      <td>LYS.SER.</td>\n",
       "      <td>HOH.LYS.SER.</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>2A5L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>2A5L-MG-2</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>K2S2</td>\n",
       "      <td>LYS.SER.</td>\n",
       "      <td>HOH.LYS.SER.</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>2A5L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>2AE8-MG-1</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>E2</td>\n",
       "      <td>GLU.</td>\n",
       "      <td>GLU.HOH.</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>2AE8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Id Element   Ion Binding family  \\\n",
       "2   5LDB-MG-1      MG  Mg²⁺             T2   \n",
       "3   2SHK-MG-1      MG  Mg²⁺         A1G2R2   \n",
       "7   2R6T-MG-1      MG  Mg²⁺           D1I1   \n",
       "8   2R6T-MG-2      MG  Mg²⁺           D1I1   \n",
       "9   2R6T-MG-3      MG  Mg²⁺           D1I1   \n",
       "10  2R6X-MG-1      MG  Mg²⁺           D1I1   \n",
       "11  2R6X-MG-2      MG  Mg²⁺           D1I1   \n",
       "12  2R6X-MG-3      MG  Mg²⁺           D1I1   \n",
       "13  2R8D-MG-1      MG  Mg²⁺             D3   \n",
       "14  2R8D-MG-2      MG  Mg²⁺             D3   \n",
       "15  2PLS-MG-1      MG  Mg²⁺             D3   \n",
       "16  2PLS-MG-2      MG  Mg²⁺             D3   \n",
       "17  2PLS-MG-3      MG  Mg²⁺             D6   \n",
       "18  2Q0Y-MG-1      MG  Mg²⁺             E2   \n",
       "19  2Q0Y-MG-2      MG  Mg²⁺             E2   \n",
       "20  2P7V-MG-1      MG  Mg²⁺           E2R2   \n",
       "21  2OX6-MG-1      MG  Mg²⁺             E2   \n",
       "22  2OX6-MG-2      MG  Mg²⁺             E2   \n",
       "23  2OX6-MG-3      MG  Mg²⁺             E2   \n",
       "24  2PAN-MG-1      MG  Mg²⁺           E2F2   \n",
       "25  2PAN-MG-2      MG  Mg²⁺           E2F2   \n",
       "26  2PC5-MG-1      MG  Mg²⁺             E3   \n",
       "30  2OI6-MG-1      MG  Mg²⁺             D3   \n",
       "31  2OIW-MG-1      MG  Mg²⁺           S2T2   \n",
       "32  2OIW-MG-2      MG  Mg²⁺           S2T2   \n",
       "33  2IEC-MG-1      MG  Mg²⁺             C2   \n",
       "34  2IEC-MG-2      MG  Mg²⁺             C2   \n",
       "35  2IOU-MG-1      MG  Mg²⁺             E3   \n",
       "36  2IOU-MG-2      MG  Mg²⁺           E3F3   \n",
       "37  2IOU-MG-3      MG  Mg²⁺           E3F3   \n",
       "38  2IOU-MG-4      MG  Mg²⁺             E3   \n",
       "39  1NC7-MG-1      MG  Mg²⁺             T4   \n",
       "40  1NC7-MG-2      MG  Mg²⁺             L4   \n",
       "41  1NC7-MG-3      MG  Mg²⁺             T4   \n",
       "42  1Q5H-MG-1      MG  Mg²⁺             E3   \n",
       "43  1Q6Z-MG-1      MG  Mg²⁺         L2N2R2   \n",
       "44  1Q6Z-MG-2      MG  Mg²⁺         L2N2R2   \n",
       "45  1Q9W-MG-1      MG  Mg²⁺             E2   \n",
       "48  1QH8-MG-1      MG  Mg²⁺         D2E1K1   \n",
       "49  1QH8-MG-2      MG  Mg²⁺         D2E1K1   \n",
       "50  1QGU-MG-1      MG  Mg²⁺         D2E1K1   \n",
       "51  1QGU-MG-2      MG  Mg²⁺         D2E1K1   \n",
       "52  1QH1-MG-1      MG  Mg²⁺         D2E1K1   \n",
       "53  1QH1-MG-2      MG  Mg²⁺         D2E1K1   \n",
       "54  2HFK-MG-1      MG  Mg²⁺             D2   \n",
       "55  2H7X-MG-1      MG  Mg²⁺             D2   \n",
       "56  2H7Y-MG-1      MG  Mg²⁺             D2   \n",
       "74  2A5L-MG-1      MG  Mg²⁺           K2S2   \n",
       "75  2A5L-MG-2      MG  Mg²⁺           K2S2   \n",
       "76  2AE8-MG-1      MG  Mg²⁺             E2   \n",
       "\n",
       "   Amino acids or nucleotide residues names All bound residues  Is homomer  \\\n",
       "2                                      THR.           HOH.THR.        True   \n",
       "3                              ALA.ARG.GLY.       ALA.GLN.HOH.        True   \n",
       "7                                  ASP.ILE.       ASP.ATP.ILE.        True   \n",
       "8                                  ASP.ILE.       ASP.ATP.ILE.        True   \n",
       "9                                  ASP.ILE.       ASP.ATP.ILE.        True   \n",
       "10                                 ASP.ILE.   ASP.ATP.HOH.ILE.        True   \n",
       "11                                 ASP.ILE.   ASP.ATP.HOH.ILE.        True   \n",
       "12                                 ASP.ILE.   ASP.ATP.HOH.ILE.        True   \n",
       "13                                     ASP.           ASP.HOH.        True   \n",
       "14                                     ASP.           ASP.HOH.        True   \n",
       "15                                     ASP.           ASP.HOH.        True   \n",
       "16                                     ASP.           ASP.HOH.        True   \n",
       "17                                     ASP.               ASP.        True   \n",
       "18                                     GLU.           GLU.HOH.        True   \n",
       "19                                     GLU.           GLU.HOH.        True   \n",
       "20                                 ARG.GLU.       ARG.GLU.HOH.        True   \n",
       "21                                     GLU.           GLU.HOH.        True   \n",
       "22                                     GLU.           GLU.HOH.        True   \n",
       "23                                     GLU.               GLU.        True   \n",
       "24                                 GLU.PHE.           GLU.PHE.        True   \n",
       "25                                 GLU.PHE.           GLU.PHE.        True   \n",
       "26                                     GLU.               GLU.        True   \n",
       "30                                     ASP.           ASP.HOH.        True   \n",
       "31                                 SER.THR.       HOH.SER.THR.        True   \n",
       "32                                 SER.THR.       HOH.SER.THR.        True   \n",
       "33                                     CYS.           CYS.HOH.        True   \n",
       "34                                     CYS.           CYS.HOH.        True   \n",
       "35                                     GLU.               GLU.        True   \n",
       "36                                 GLU.PHE.           GLU.PHE.        True   \n",
       "37                                 GLU.PHE.           GLU.PHE.        True   \n",
       "38                                     GLU.               GLU.        True   \n",
       "39                                     THR.               THR.        True   \n",
       "40                                     LEU.               LEU.        True   \n",
       "41                                     THR.               HOH.        True   \n",
       "42                                     GLU.           GLU.HOH.        True   \n",
       "43                             ARG.ASN.LEU.       ARG.ASN.LEU.        True   \n",
       "44                             ARG.ASN.LEU.       ARG.ASN.LEU.        True   \n",
       "45                                     GLU.           GLU.HOH.        True   \n",
       "48                             ASP.GLU.LYS.               HOH.        True   \n",
       "49                             ASP.GLU.LYS.   ASP.GLU.HOH.LYS.        True   \n",
       "50                             ASP.GLU.LYS.               HOH.        True   \n",
       "51                             ASP.GLU.LYS.   ASP.GLU.HOH.LYS.        True   \n",
       "52                             ASP.GLU.LYS.               HOH.        True   \n",
       "53                             ASP.GLU.LYS.   ASP.GLU.HOH.LYS.        True   \n",
       "54                                     ASP.           ASP.HOH.        True   \n",
       "55                                     ASP.           ASP.HOH.        True   \n",
       "56                                     ASP.           ASP.HOH.        True   \n",
       "74                                 LYS.SER.       HOH.LYS.SER.        True   \n",
       "75                                 LYS.SER.       HOH.LYS.SER.        True   \n",
       "76                                     GLU.           GLU.HOH.        True   \n",
       "\n",
       "    Number of bound amino acids or nucleotide residues  \\\n",
       "2                                                   2    \n",
       "3                                                   5    \n",
       "7                                                   2    \n",
       "8                                                   2    \n",
       "9                                                   2    \n",
       "10                                                  2    \n",
       "11                                                  2    \n",
       "12                                                  2    \n",
       "13                                                  3    \n",
       "14                                                  3    \n",
       "15                                                  3    \n",
       "16                                                  3    \n",
       "17                                                  6    \n",
       "18                                                  2    \n",
       "19                                                  2    \n",
       "20                                                  4    \n",
       "21                                                  2    \n",
       "22                                                  2    \n",
       "23                                                  2    \n",
       "24                                                  4    \n",
       "25                                                  4    \n",
       "26                                                  3    \n",
       "30                                                  3    \n",
       "31                                                  4    \n",
       "32                                                  4    \n",
       "33                                                  2    \n",
       "34                                                  2    \n",
       "35                                                  3    \n",
       "36                                                  6    \n",
       "37                                                  6    \n",
       "38                                                  3    \n",
       "39                                                  4    \n",
       "40                                                  4    \n",
       "41                                                  4    \n",
       "42                                                  3    \n",
       "43                                                  6    \n",
       "44                                                  6    \n",
       "45                                                  2    \n",
       "48                                                  4    \n",
       "49                                                  4    \n",
       "50                                                  4    \n",
       "51                                                  4    \n",
       "52                                                  4    \n",
       "53                                                  4    \n",
       "54                                                  2    \n",
       "55                                                  2    \n",
       "56                                                  2    \n",
       "74                                                  4    \n",
       "75                                                  4    \n",
       "76                                                  2    \n",
       "\n",
       "    Number of all bound residues  Number of coordinating chains  \\\n",
       "2                              3                              2   \n",
       "3                              5                              2   \n",
       "7                              3                              2   \n",
       "8                              3                              2   \n",
       "9                              3                              2   \n",
       "10                             4                              2   \n",
       "11                             4                              2   \n",
       "12                             4                              2   \n",
       "13                             6                              2   \n",
       "14                             6                              2   \n",
       "15                             6                              2   \n",
       "16                             6                              2   \n",
       "17                             6                              2   \n",
       "18                             5                              2   \n",
       "19                             5                              2   \n",
       "20                             6                              2   \n",
       "21                             3                              2   \n",
       "22                             4                              2   \n",
       "23                             2                              2   \n",
       "24                             4                              2   \n",
       "25                             4                              2   \n",
       "26                             3                              3   \n",
       "30                             6                              3   \n",
       "31                             6                              2   \n",
       "32                             6                              2   \n",
       "33                             6                              2   \n",
       "34                             5                              2   \n",
       "35                             3                              3   \n",
       "36                             6                              3   \n",
       "37                             6                              3   \n",
       "38                             3                              3   \n",
       "39                             4                              4   \n",
       "40                             4                              4   \n",
       "41                             2                              4   \n",
       "42                             4                              3   \n",
       "43                             6                              2   \n",
       "44                             6                              2   \n",
       "45                             6                              2   \n",
       "48                             6                              2   \n",
       "49                             6                              2   \n",
       "50                             6                              2   \n",
       "51                             6                              2   \n",
       "52                             6                              2   \n",
       "53                             6                              2   \n",
       "54                             4                              2   \n",
       "55                             4                              2   \n",
       "56                             4                              2   \n",
       "74                             6                              2   \n",
       "75                             6                              2   \n",
       "76                             4                              2   \n",
       "\n",
       "    Representative pdb_id  \n",
       "2             True   5LDB  \n",
       "3            False   2SHK  \n",
       "7            False   2R6T  \n",
       "8            False   2R6T  \n",
       "9            False   2R6T  \n",
       "10           False   2R6X  \n",
       "11           False   2R6X  \n",
       "12           False   2R6X  \n",
       "13            True   2R8D  \n",
       "14           False   2R8D  \n",
       "15            True   2PLS  \n",
       "16            True   2PLS  \n",
       "17            True   2PLS  \n",
       "18            True   2Q0Y  \n",
       "19           False   2Q0Y  \n",
       "20            True   2P7V  \n",
       "21            True   2OX6  \n",
       "22           False   2OX6  \n",
       "23           False   2OX6  \n",
       "24            True   2PAN  \n",
       "25           False   2PAN  \n",
       "26           False   2PC5  \n",
       "30           False   2OI6  \n",
       "31            True   2OIW  \n",
       "32           False   2OIW  \n",
       "33            True   2IEC  \n",
       "34           False   2IEC  \n",
       "35           False   2IOU  \n",
       "36           False   2IOU  \n",
       "37           False   2IOU  \n",
       "38           False   2IOU  \n",
       "39            True   1NC7  \n",
       "40            True   1NC7  \n",
       "41            True   1NC7  \n",
       "42            True   1Q5H  \n",
       "43            True   1Q6Z  \n",
       "44           False   1Q6Z  \n",
       "45            True   1Q9W  \n",
       "48           False   1QH8  \n",
       "49           False   1QH8  \n",
       "50            True   1QGU  \n",
       "51           False   1QGU  \n",
       "52           False   1QH1  \n",
       "53           False   1QH1  \n",
       "54            True   2HFK  \n",
       "55           False   2H7X  \n",
       "56           False   2H7Y  \n",
       "74            True   2A5L  \n",
       "75           False   2A5L  \n",
       "76            True   2AE8  "
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_homo.head(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#df_homo['pdb_id'].str.contains('4HHB').sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_homo.to_csv('df_homomer.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [],
   "source": [
    "from urllib.request import urlretrieve\n",
    "import requests\n",
    "from tqdm import tqdm\n",
    "import validators"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('datasets/rcsb_pdbs/2AE8.pdb', <http.client.HTTPMessage at 0x7f4aa1a3b9a0>)"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "url = 'https://files.rcsb.org/download/2AE8.pdb'\n",
    "file = 'datasets/rcsb_pdbs/2AE8.pdb'\n",
    "urlretrieve(url, file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  4%|▍         | 138/3199 [01:05<23:28,  2.17it/s] "
     ]
    }
   ],
   "source": [
    "logs = {'pdb_id':[], 'status':[]}\n",
    "for pdb_id in tqdm(df_homo['pdb_id'].unique()):\n",
    "    logs['pdb_id'].append(pdb_id)\n",
    "    url = f'https://files.rcsb.org/download/{pdb_id}.pdb'\n",
    "    file = f'datasets/rcsb_pdbs/{pdb_id}.pdb'\n",
    "    r = requests.get(url)\n",
    "    if '404' in str(r):\n",
    "        logs['status'].append(0)\n",
    "        continue\n",
    "    urlretrieve(url,file)\n",
    "    logs['status'].append(1)\n",
    "logs = pd.DataFrame(logs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "logs.to_csv('datasets.logs')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #pdb_id\n",
    "# validators.url(\"https://files.rcsb.org/download/4QYK.pdb\")\n",
    "# url = \"https://files.rcsb.org/download/4QYK.pdb\"\n",
    "# url = \"https://files.rcsb.org/download/2IEC.pdb\"\n",
    "# requests.get(url)\n",
    "# r = requests.get(url)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### analyze datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "logs = pd.read_csv('datasets.logs')#, index_col=False)\n",
    "df_homo = pd.read_csv('df_homomer.csv')\n",
    "df = pd.read_csv('table.csv')\n",
    "#pdb_ok = logs[logs['status']==1]['pdb_id']\n",
    "#df_homo_filter = df_homo[df_homo['pdb_id'].isin(pdb_ok)]\n",
    "df_homo_filter = pd.read_csv('df_homo_filter.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_homo_filter['metal'] = df_homo_filter['Id'].str.split('-', expand=True)[1]\n",
    "# df_homo_filter.to_csv('df_homo_filter.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_homo_filter['metal'].isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_homo_filter.to_csv('df_homo_filter.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pdb_ok.str.contains('8FJD')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#drop row in df_homo if pdb_id is not in pdb_ok"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3132"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_homo_filter['pdb_id'].nunique()\n",
    "#df_homo_filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_homo[df_homo['pdb_id'] =='7PGI']\n",
    "# df[df['Id'].str.contains('7PGI')]['Amino acids or nucleotide residues names'].str.split('.')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['31', 'B', 'NI', array([-2.000e-03, -3.000e-03, -8.328e+00])],\n",
       " ['32', 'B', 'NI', array([  7.202,   6.801, -19.948])],\n",
       " ['31', 'D', 'NI', array([ 0.035, -0.117,  8.724])],\n",
       " ['32', 'D', 'NI', array([-9.533,  2.855, 20.337])]]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "util.get_ligand_num('datasets/rcsb_pdbs/2R36.pdb', 'NI', multi=True)#[0][1]#.split(',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hetxyz [[-2.000e-03 -3.000e-03 -8.328e+00]]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[['10', 'B', 'HIS']]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "util.get_ligand_contacts('datasets/rcsb_pdbs/2R36.pdb',hetatm_num=31, hetatm='NI',hetatm_chain='B', cutoff=8.0, hotspot=['HIS'])\n",
    "#'NI' in 'NI'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_homo_filter['metal_binding_id'] = df_homo_filter['Id'].str.split('-', expand=True)[2]\n",
    "# df_homo_filter.to_csv('df_homo_filter.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_homo_filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# [int(i) for i in df_homo_filter['Ion'].str.split()]#.[int(i) for i in test_string.split() if i.isdigit()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "id_debug = '4XLE'\n",
    "debug = False\n",
    "#for i,row in df_homo_filter.iterrows():\n",
    "for pdb_id in df_homo_filter['pdb_id'].unique():\n",
    "    if pdb_id != id_debug and debug:\n",
    "        continue\n",
    "    sample_df = df_homo_filter[df_homo_filter['pdb_id'] == pdb_id]\n",
    "    row = sample_df[sample_df['Representative'] == 1].head(1)\n",
    "    if len(row) == 0:\n",
    "        row = sample_df.sort_values(by='metal_binding_id').head(1)\n",
    "    hotspots_raw = row['Amino acids or nucleotide residues names'].values[0].split('.')\n",
    "    hotspots_raw.remove('')\n",
    "    hotspots = []\n",
    "    for h in hotspots_raw:\n",
    "        if h in util.THREE_TO_ONE.keys():\n",
    "            hotspots.append(h)\n",
    "    metal_name = row['metal'].values[0]\n",
    "    metal_binding_id = row['metal_binding_id'].values[0]\n",
    "    no_bound_residues = row['Number of bound amino acids or nucleotide residues'].values[0]\n",
    "    motif_residue = []\n",
    "\n",
    "    if hotspots==[]:\n",
    "        continue #sample more cause oversampling, or non-canonical resi\n",
    "    metal_infos = util.get_ligand_num(f'datasets/rcsb_pdbs/{pdb_id}.pdb', metal_name, multi=True)\n",
    "    for metal_info in metal_infos:\n",
    "        metal_num, metal_chain, metal_name, metal_xyz = metal_info\n",
    "        motif = util.get_ligand_contacts(f'datasets/rcsb_pdbs/{pdb_id}.pdb', hetatm_num=metal_num, hetatm=metal_name,hetatm_chain=metal_chain, cutoff=6.0,hotspot=hotspots) # motif_resi_num, motif_chain, motif_resi_id\n",
    "        motif_residue = [item[-1] for item in motif]\n",
    "        if len(motif) >=  len(hotspots):# and :\n",
    "            if set(hotspots) == set(motif_residue):\n",
    "                break\n",
    "            elif len(set(motif_residue)) == len(hotspots):\n",
    "                break\n",
    "\n",
    "    with open(f'datasets/rcsb_pdbs/{pdb_id}.pdb', 'r') as f:\n",
    "        w_lines = []\n",
    "        for line in f.readlines():\n",
    "            if line.startswith('HETATM'):\n",
    "                if line[17:21].strip() == metal_name and line[22:26].strip() == metal_num and line[21:22].strip() == metal_chain:\n",
    "                    w_lines.append(line[:30] + '{:8.3f}'.format(0.0) + '{:8.3f}'.format(0.0) + '{:8.3f}'.format(0.0) + line[54:])\n",
    "            if line.startswith('ATOM'):\n",
    "                for m in motif:\n",
    "                    #[line[22:26].strip(), str(line[21:22].strip()),line[17:21].strip()])\n",
    "                    if line[22:26].strip()== m[0] and line[21:22].strip() == m[1] and line[17:21].strip()== m[2]:\n",
    "                        line_xyz = np.array([float(line[30:38]), float(line[38:46]), float(line[46:54])])\n",
    "                        w_lines.append(line[:30] + '{:8.3f}'.format(line_xyz[0]-metal_xyz[0]) + '{:8.3f}'.format(line_xyz[1]-metal_xyz[1]) + '{:8.3f}'.format(line_xyz[2]-metal_xyz[2]) + line[54:])\n",
    "\n",
    "    with open(f'datasets/rotamer_motif/{pdb_id}_{metal_binding_id}.pdb', 'w') as f1:\n",
    "        f1.writelines(w_lines)\n",
    "    if pdb_id == id_debug and debug:\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Id</th>\n",
       "      <th>Element</th>\n",
       "      <th>Ion</th>\n",
       "      <th>Binding family</th>\n",
       "      <th>Amino acids or nucleotide residues names</th>\n",
       "      <th>All bound residues</th>\n",
       "      <th>Is homomer</th>\n",
       "      <th>Number of bound amino acids or nucleotide residues</th>\n",
       "      <th>Number of all bound residues</th>\n",
       "      <th>Number of coordinating chains</th>\n",
       "      <th>Representative</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>metal_binding_id</th>\n",
       "      <th>metal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3738</th>\n",
       "      <td>4XLE-NA-3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Na⁺</td>\n",
       "      <td>N1T1V1</td>\n",
       "      <td>ASN.THR.VAL.</td>\n",
       "      <td>ASN.HOH.THR.VAL.</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>4XLE</td>\n",
       "      <td>3</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             Id Element  Ion Binding family  \\\n",
       "3738  4XLE-NA-3     NaN  Na⁺         N1T1V1   \n",
       "\n",
       "     Amino acids or nucleotide residues names All bound residues  Is homomer  \\\n",
       "3738                             ASN.THR.VAL.   ASN.HOH.THR.VAL.        True   \n",
       "\n",
       "      Number of bound amino acids or nucleotide residues  \\\n",
       "3738                                                  3    \n",
       "\n",
       "      Number of all bound residues  Number of coordinating chains  \\\n",
       "3738                             6                              2   \n",
       "\n",
       "      Representative pdb_id  metal_binding_id metal  \n",
       "3738            True   4XLE                 3    NA  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_homo_filter[df_homo_filter['pdb_id'] == '1AEW'].sort_values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3, [], ['143', 'A', 'HIS'], [])"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# bound_residues, pdb_id, metal_name\n",
    "no_bound_residues, motif, m, motif_residue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# metal_infos= [['401', 'A', 'MG', np.array([126.878, 123.014,  36.517])],\n",
    "#   ['401', 'B', 'MG', np.array([136.216,  88.583,  29.722])],\n",
    "#   ['401', 'C', 'MG', np.array([136.062,  95.724,  -5.958])],\n",
    "#   ['401', 'D', 'MG', np.array([126.643, 130.136,   0.986])],\n",
    "#   ['402', 'D', 'MG', np.array([121.658, 118.475,  -8.37 ])],\n",
    "#   ['401', 'E', 'MG', np.array([ 30.586, 119.339,  -7.67 ])],\n",
    "#   ['401', 'F', 'MG', np.array([ 26.262, 129.272,  26.918])],\n",
    "#   ['401', 'G', 'MG', np.array([34.604, 95.578, 37.712])],\n",
    "#   ['401', 'H', 'MG', np.array([38.928, 85.574,  3.026])]]\n",
    "# pdb_id= '3N6Q'\n",
    "# hotspots = ['GLU']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "([['1010', 'A', 'NA', array([49.953, 26.317, 55.249])],\n",
       "  ['1011', 'A', 'NA', array([27.026, 50.465, 68.139])],\n",
       "  ['1012', 'A', 'NA', array([26.147, 53.467, 24.209])]],\n",
       " '4XLE')"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# pdb_id= '3H9F'\n",
    "# metal_name = 'MG'\n",
    "metal_infos = util.get_ligand_num(f'datasets/rcsb_pdbs/{pdb_id}.pdb', metal_name, multi=True)\n",
    "metal_infos, pdb_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(['ASN', 'THR', 'VAL'], [])"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# pdb_id, metal_infos\n",
    "hotspots, motif_residue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1010 A NA [49.953 26.317 55.249]\n",
      "[['516', 'A', 'ASN'], ['517', 'A', 'ASN'], ['545', 'A', 'VAL']]\n",
      "['ASN', 'ASN', 'VAL']\n",
      "1011 A NA [27.026 50.465 68.139]\n",
      "[['534', 'A', 'VAL'], ['567', 'A', 'VAL']]\n",
      "['VAL', 'VAL']\n",
      "1012 A NA [26.147 53.467 24.209]\n",
      "[['249', 'A', 'ASN']]\n",
      "['ASN']\n"
     ]
    }
   ],
   "source": [
    "motif_residue = []\n",
    "for metal_info in metal_infos:\n",
    "    metal_num, metal_chain, metal_name, metal_xyz = metal_info\n",
    "    print(metal_num, metal_chain, metal_name, metal_xyz)\n",
    "    motif = util.get_ligand_contacts(f'datasets/rcsb_pdbs/{pdb_id}.pdb', hetatm_num=metal_num, hetatm=metal_name,hetatm_chain=metal_chain, cutoff=8.0,hotspot=hotspots) # motif_resi_num, \n",
    "    motif_residue = [item[-1] for item in motif]\n",
    "    print(motif)\n",
    "    print(motif_residue)\n",
    "    #     break\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # metal_infos, \n",
    "# motif = util.get_ligand_contacts(f'datasets/rcsb_pdbs/{pdb_id}.pdb', hetatm_num=metal_num, hetatm=metal_name,hetatm_chain=metal_chain, cutoff=5.0,hotspot=hotspots)\n",
    "# len(motif)\n",
    "# motif"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# only get last element of motif\n",
    "motif\n",
    "motif"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "hetatm_num=metal_num\n",
    "hetatm=metal_name\n",
    "hetatm_chain=metal_chain\n",
    "cutoff=5.0\n",
    "ca_xyz = np.zeros(3)\n",
    "cb_xyz = np.zeros(3)\n",
    "hetatm_xyz = np.zeros(3)\n",
    "# chain = chain.split(',')\n",
    "# chain = [i.strip().upper() for i in chain]\n",
    "with open(f'datasets/rcsb_pdbs/{pdb_id}.pdb', 'r') as f:\n",
    "    for line in f.readlines():\n",
    "        if line.startswith('ATOM'):\n",
    "            # if line[21:22].strip() in chain:\n",
    "            if line[13:15] =='CA':\n",
    "                ca_xyz = np.vstack((ca_xyz,np.array([float(line[30:38].strip()), float(line[38:46].strip()), float(line[46:54].strip())])))\n",
    "            elif line[13:15] =='CB':\n",
    "                cb_xyz = np.vstack((cb_xyz,np.array([float(line[30:38].strip()), float(line[38:46].strip()), float(line[46:54].strip())])))\n",
    "        elif line.startswith('HETATM'):\n",
    "            \n",
    "            if (line[17:21].strip() in hetatm) and ((line[22:26].strip()) == hetatm_num) and (line[21:22].strip() in hetatm_chain):\n",
    "                \n",
    "                hetatm_xyz = np.vstack((hetatm_xyz,np.array([float(line[30:38].strip()), float(line[38:46].strip()), float(line[46:54].strip())])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[ 0.   ,  0.   ,  0.   ],\n",
       "       [38.928, 85.574,  3.026]])"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hetatm_xyz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "pdb_id, metal_binding_id\n",
    "motif = util.get_ligand_num(f'datasets/rcsb_pdbs/{pdb_id}.pdb', metal_name, multi=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Id</th>\n",
       "      <th>Element</th>\n",
       "      <th>Ion</th>\n",
       "      <th>Binding family</th>\n",
       "      <th>Amino acids or nucleotide residues names</th>\n",
       "      <th>All bound residues</th>\n",
       "      <th>Is homomer</th>\n",
       "      <th>Number of bound amino acids or nucleotide residues</th>\n",
       "      <th>Number of all bound residues</th>\n",
       "      <th>Number of coordinating chains</th>\n",
       "      <th>Representative</th>\n",
       "      <th>pdb_id</th>\n",
       "      <th>metal_binding_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2R8D-MG-1</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>D3</td>\n",
       "      <td>ASP.</td>\n",
       "      <td>ASP.HOH.</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>2R8D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2R8D-MG-2</td>\n",
       "      <td>MG</td>\n",
       "      <td>Mg²⁺</td>\n",
       "      <td>D3</td>\n",
       "      <td>ASP.</td>\n",
       "      <td>ASP.HOH.</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>2R8D</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Id Element   Ion Binding family  \\\n",
       "8  2R8D-MG-1      MG  Mg²⁺             D3   \n",
       "9  2R8D-MG-2      MG  Mg²⁺             D3   \n",
       "\n",
       "  Amino acids or nucleotide residues names All bound residues  Is homomer  \\\n",
       "8                                     ASP.           ASP.HOH.        True   \n",
       "9                                     ASP.           ASP.HOH.        True   \n",
       "\n",
       "   Number of bound amino acids or nucleotide residues  \\\n",
       "8                                                  3    \n",
       "9                                                  3    \n",
       "\n",
       "   Number of all bound residues  Number of coordinating chains  \\\n",
       "8                             6                              2   \n",
       "9                             6                              2   \n",
       "\n",
       "   Representative pdb_id  metal_binding_id  \n",
       "8            True   2R8D                 1  \n",
       "9           False   2R8D                 2  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_homo_filter[df_homo_filter['pdb_id'] == '2R8D']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "12503"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(df_homo_filter['Representative'] == 0).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "#(df_homo_filter['Representative']==1).sum()\n",
    "count = 0 \n",
    "for pdb_id in (df_homo_filter['pdb_id'].unique()):\n",
    "    if sum(df_homo_filter[(df_homo_filter['pdb_id'] == pdb_id)]['Representative']==1)==0:\n",
    "        count+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1294"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "count"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SE3nv_dimaio",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
